pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,358 @@
+}
+
+func getTeamByID(client *cherrygo.Client, ID string) *cherrygo.Teams {
+	teams, _, err := client.Teams.List()
+	if err != nil {
+		return nil
+	}
+
+	for _, team := range teams {
+		if string(team.ID) == ID {
+			return &team
+		}
+	}
+	return nil
+}
+
+func getPlanByName(client *cherrygo.Client, planName string, teamID string) *cherrygo.Plans {
+	tid, err := strconv.Atoi(teamID)
+	if err != nil {
+		return nil
+	}
+
+	plans, _, err := client.Plans.List(tid)
+	if err != nil {
+		return nil
+	}
+
+	for _, plan := range plans {
+		if plan.Name == planName {
+			return &plan
+		}
+	}
+	return nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.Token == "" {
+		return errors.New("token is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.Token)
+
+	if c.TeamID != "" {
+		if getTeamByID(client, c.TeamID) == nil {
+			return fmt.Errorf("failed to get teamID")
+		}
+	}
+
+	if c.Plan != "" {
+		if getPlanByName(client, c.Plan, c.TeamID) == nil {
+			return fmt.Errorf("failed to get plan")
+		}
+	}
+
+	if c.Location != "" {
+		if c.Location != "EU-East-1" && c.Location != "EU-West-1" {
+			return fmt.Errorf("failed to get location")
+		}
+	}
+
+	if c.ProjectID != "" {
+		if _, _, err := client.Project.List(c.ProjectID); err != nil {
+			return fmt.Errorf("failed to get projectID")
+		}
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	osName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	teamID, err := strconv.Atoi(c.TeamID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid TeamID, details = %v", err),
+		}
+	}
+
+	var planID int
+	plans, _, err := client.Plans.List(teamID)
+	for _, plan := range plans {
+		if plan.Name == c.Plan {
+			planID = plan.ID
+			break
+		}
+	}
+
+	serverCreateRequest := cherrygo.CreateServer{
+		ProjectID:   c.ProjectID,
+		Hostname:    machine.Name,
+		Image:       osName,
+		Region:      c.Location,
+		SSHKeys:     []string{},
```

Comment:
Do we need to specify `SSHKeys` and `IPAddresses` here if those are empty?

Reply:
Empty fields removed